README
Catalyst-Controller-POD

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Catalyst::Controller::POD

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Catalyst-Controller-POD

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Catalyst-Controller-POD

    CPAN Ratings
        http://cpanratings.perl.org/d/Catalyst-Controller-POD

    Search CPAN
        http://search.cpan.org/dist/Catalyst-Controller-POD


COPYRIGHT AND LICENCE

Copyright (C) 2008 Moritz Onken

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Revision history for Catalyst-Controller-POD

0.02007  2010-03-07
         - Added additional configuration options:
           expanded_module_tree, home_tab_content, initial_module,
           show_home_tab (Tristan Pratt)
         - Added permalinks (Tristan Pratt)
         - Update to ExtJS 3.1.1 (fixes memory leaks in browsers)

0.02006  2009-11-20
         - fixed another warning (#51681)

0.02005  2009-11-20
         - use next instead of NEXT
         - Added "not found" page for namespaces like DBIx

0.02004  2009-05-04
         - JS cleanups
         - filtered results do not expand by default
         - moved to git repository

0.02003  2009-05-03
         - Works with Catalyst 5.8 (reported by T. Hosaka)

0.02002  2008-08-16
         - requires parent
         - ext cleanup

0.02001  2008-08-09
         - removed Pod::TOC dependency

0.01_03  2008-07-30
         - A lot of new feature
         - added some basic tests, more to come

0.01_01    Date/time
        First version, released on an unsuspecting world.



t/lib/TestApp.pm
package TestApp;

use strict;
use warnings;

use Catalyst::Runtime '5.70';

use parent qw/Catalyst/;

our $VERSION = '0.01';
__PACKAGE__->config( name => 'TestApp' ,
"Controller::Root" => {}
);
__PACKAGE__->setup(qw//);


1;


t/lib/TestApp/Controller/Docs.pm
package TestApp::Controller::Docs;

use strict;
use warnings;
use base 'Catalyst::Controller::POD';


sub test : Local {
    my ( $self, $c ) = @_;

    $c->response->body( "here I am" );
}


1;


t/lib/TestApp/Controller/Root.pm
package TestApp::Controller::Root;

use strict;
use warnings;
use base 'Catalyst::Controller::POD';

__PACKAGE__->config->{namespace} = '';

sub test : Local {
    my ( $self, $c ) = @_;

    $c->response->body( "here I am" );
}


1;


Makefile.PL
use inc::Module::Install;

# Define metadata
name 'Catalyst-Controller-POD';
all_from 'lib/Catalyst/Controller/POD.pm';

repository 'http://github.com/monken/Catalyst-Controller-POD/tree/master';

# Specific dependencies
requires 'Test::More'            => 0;
requires 'Pod::Simple'           => 0;
requires 'File::Find'            => "0.05";
requires 'File::ShareDir'        => "0.05";
requires 'File::Slurp'           => "9999";
requires 'Pod::Simple'           => "3.05";
#requires 'File::Wildcard';
requires 'JSON::XS'              => "2.21";
requires 'Pod::POM'              => "0.17";
requires 'LWP::Simple'           => "5.810";
requires "XML::Simple"           => "2.18";
requires 'List::MoreUtils'       => "0.22";
requires 'Class::Accessor::Fast' => "0.31";
requires 'Catalyst::Controller'  => 0;
requires 'Pod::POM::View::TOC'   => "0.02";
requires 'parent'                => 0;
requires 'Path::Class::File' => 0;

test_requires 'Test::More'     => '0.42';
test_requires 'Test::WWW::Mechanize::Catalyst' => "0.41";

install_share;

WriteAll;


share/docs.js
Ext.namespace("POD");

// Too much regex for my liking.  A JQuery plugin would be cleaner, but this works.
// http://stackoverflow.com/questions/901115/get-querystring-with-jquery/901144#901144
POD.qsVal = function ( name ) {
  name = name.replace(/[\[]/,"\\\[").replace(/[\]]/,"\\\]");
  var regexS = "[\\?&]"+name+"=([^&#]*)";
  var regex = new RegExp( regexS );
  var results = regex.exec( window.location.href );
  if( results == null )
    return "";
  else
    return results[1];
}

POD.addTab = function (title, url, section,module) {
	if(Ext.getCmp("tab-"+title)) {
		Ext.getCmp("tab-"+title).show();
		return Ext.getCmp("tab-"+title);
	}
	POD.clearTOC();
	var tab = POD.tabs.add({
		title: title, 
		id: "tab-" + title,
		closable: true,
		closeAction: "hide",
		autoLoad: {url : url,
		loadScripts: true,
		callback : function() {POD.scrollToSection(tab, section,module);}},
		autoScroll: true
		}
	);
	tab.show();
	return tab;
	
}

POD.proxyLink = function(link) {
	var re = new RegExp("^" + Ext.escapeRe("[% root %]"));
	if(!re.test(link.href)) return true;
	var module = link.href.replace(new RegExp(Ext.escapeRe("[% root %]/module/")), "");
	
		var parts = module.split(/#/);
		section = unescape(parts[1]);
		module = parts[0];
		if(re.test(module)) {
			module = POD.tabs.getActiveTab().getId().replace(/tab-/, "");
		}
	var tab = POD.addTab(module, link.href, section, module);
	POD.scrollToSection(tab, section, module);
	return false;
	
}

POD.scrollToSection = function(tab, section, module){
	var el = document.getElementById("section-"+module+"-"+section);
	if(el){
		var top = (Ext.fly(el).getOffsetsTo(tab.body)[1]) + tab.body.dom.scrollTop;
		tab.body.scrollTo('top', top, {duration:.5});
    }
}

POD.reloadTree = function(e) {
	if(!e.target.value) {
		POD.tree.getLoader().dataUrl = '[% root %]/modules';
		POD.tree.getLoader().load(POD.tree.root);
		return;
	}
	POD.tree.getLoader().dataUrl = '[% root %]/modules/'+e.target.value;
	POD.tree.getLoader().load(POD.tree.root);
}

POD.tree = new Ext.tree.TreePanel({
	title: "Modules",
	autoScroll:true,
	// collapsible:true,
	animate:true,
	rootVisible: false,
	split: true,
	region: "center",
containerScroll: true,
listeners:  {
	"click" : function(node) { POD.addTab(node.attributes.name, "[% root %]/module/"+node.attributes.name) }},
loader: new Ext.tree.TreeLoader({
	dataUrl:'[% root %]/modules',
	autoLoad: false,
	listeners: { 
		'beforeload' : function() { POD.tree.getEl().mask('Modules are being loaded') }, 
		'load' : function() { 
			POD.tree.getEl().unmask();
			if([% expand_module_tree_on_load %]) {
				var old = POD.tree.animate;
				POD.tree.animate = false;
				POD.tree.expandAll();
				POD.tree.animate = old;
			}
		}
	}

}),
tbar: [new Ext.form.TextField({
	width: 130,
	emptyText:'Find a Class',
	listeners:{
		render: function(f){
			f.el.on('keydown', POD.reloadTree, f, {buffer: 350});
		}
	}
}), ' ', ' ',{
 	   handler: function (){POD.tree.expandAll()},
 	   tooltip: 'Expand all nodes',
 	   iconCls:"icon-expand-all"},
 	  {
 	 	   handler: function (){POD.tree.collapseAll()},
 	 	   tooltip: 'Collapse all nodes',
 	 	   iconCls:"icon-collapse-all"}
 ]
}) ;

var root = new Ext.tree.AsyncTreeNode({
	text: 'mods',
	expanded:true,
	// id:'root',
});

POD.tree.setRootNode(root);
POD.filter = new Ext.tree.TreeFilter(POD.tree, {
	clearBlank: true,
	autoClear: true
});

POD.TOC = new Ext.tree.TreePanel({
	title: "TOC",
	autoScroll:true,
	height: 200,
	collapsible:true,
	animate:true,
	rootVisible: false,
	loader: new Ext.tree.TreeLoader(),
	split: true,
	region: "north",
listeners:  {
	"click" : function(node) { 
	  var module = POD.tabs.getActiveTab().getId().replace(/tab-/,"");
	  // alert(node.text + module);
	  POD.scrollToSection(POD.tabs.getActiveTab(), node.text, module);
}}
}) ;
POD.TOC.setRootNode(new Ext.tree.AsyncTreeNode({
	text: 'mods',
	expanded:false
}));

POD.populateTOC = function (nodes, root) {
	for(var i = 0; i < nodes.length; i++) {
		var n = POD.TOC.getLoader().createNode(nodes[i]);
		if (n) {
			root.appendChild(n);
		}
	}
}

POD.leftColumn = new Ext.Panel({
	layout: "border",
	region: "west",
	split: true,
	width: 200,
	items: [POD.tree, POD.TOC]
	
});

POD.TOCs = {};

POD.clearTOC = function() {
	while (POD.TOC.root.firstChild) {
		POD.TOC.root.removeChild(POD.TOC.root.firstChild);
	}
}

POD.setTOC = function (nodes) {
	POD.clearTOC();
	POD.TOCs[POD.tabs.getActiveTab().getId()] = nodes;
    POD.populateTOC(nodes, POD.TOC.root);
}


POD.searchStore = new Ext.data.Store( {
	proxy :new Ext.data.HttpProxy( {
		method :'POST',
		url :'[% root %]/search'
	}),
	reader :new Ext.data.JsonReader( {
		root :'module',
		totalProperty :'matches'
	}, [ {
		name :'link',
		mapping :'link'
	}, {
		name :'name',
		mapping :'name'
	}, {
		name :'released',
		mapping :'released'
	}, {
		name :'version',
		mapping :'version'
	}, {
		name :'description',
		mapping :'description'
	}, {
		name :'author',
		mapping :'author'
	} ])
});

var resultTpl = new Ext.XTemplate(
		'<tpl for="."><div class="search-item">',
		'<span>{released}',
		'</span><h3>{name}</h3> (Version {version})<br>{description}</div></tpl>');


var Memoria = {};
Memoria.Search = {};
Memoria.Search.InstantAdd = new (function(){
this.init = function(combo) {
	this.combo = combo;
	var add = new Ext.Layer({cls: "x-combo-list", html: "test1234", zindex: "100"});	
    var lw = combo.listWidth || Math.max(combo.wrap.getWidth(), combo.minListWidth);
    add.setWidth(lw);
	add.alignTo(combo.wrap, combo.listAlign);
	this.content = add.createChild({cls:'search-item'});
	this.add = add;
	new Ext.KeyNav(combo.el, {
		"esc": function() { Memoria.Search.InstantAdd.hide() },
		"enter": function() {
			Memoria.Search.InstantAdd.hide();
			Memoria.Clients.add(stringToName(Memoria.Search.InstantAdd.combo.getValue()));
		}
	})
	combo.el.on("keydown", this.hide , this);
	this.tpl = new Ext.XTemplate('<h3 style="text-align: center">Sorry. Couldn\'t find anything.</h3>');
}

this.show = function () {
	if(!this.combo.getValue()) return;
	this.add.show();
	this.content.dom.innerHTML = this.tpl.apply({name: this.combo.getValue()});
}

this.hide = function () {
	this.add.dom.style.visibility=""; // Do not use this.add.hide() BUG?
}
})();


POD.tabs = 
    new Ext.TabPanel({
        region:'center',
        activeTab:0,
        autoScroll: true,
        margins: "5 5 5 5",
        enableTabScroll: true,
        listeners: {"tabchange": function(panel, tab) {
    		if(POD.TOCs[tab.getId()])
    			POD.setTOC(POD.TOCs[tab.getId()]);
    		if(tab.getId() == "search-box")
    			POD.clearTOC();
    	}},
        tools: [{id: "print", handler: function() { window.open("[% root %]/module/"+tabs.getActiveTab().id.replace(/tab-/,"")); }},
                {id: "close",handler: function (){ tabs.items.each(function(el){if(new RegExp("tab-").test(el.id)) tabs.remove(el)}) }}],
    });

//
// For some reason the following order matters.
// 

// Display documentation for the permalink
var permalink = POD.qsVal("permalink");
if(permalink) {
	POD.addTab(permalink, "[% root %]/module/" + permalink);
}

// Show the home tab only if we're configured to 
if([% show_home_tab %]) {
	POD.tabs.add({
		layout :'form',
		title : "Home",
		id : "search-box",
		frame :false,
		border :false,
		autoLoad: {
			url : "[% root %]/home_tab_content",
			callback : function() { POD.configSearchCombo(); },
			loadScripts: true,
		},
	});
}

if("[% initial_module %]") {
	// Display a perldoc tab on startup
	POD.addTab("[% initial_module %]", "[% root %]/module/[% initial_module %]");
}

POD.configSearchCombo = function () {
    var search = new Ext.form.ComboBox( {
    	store :POD.searchStore,
    	typeAhead :false,
    	minChars :3,
    	queryParam :'value',
    	emptyText: "Search the CPAN for modules",
    	loadingText :'Searching ...',
    	width :470,
    	pageSize :50,
    	hideTrigger :true,
    	tpl :resultTpl,
    	applyTo :Ext.getDom('search'),
    	itemSelector :'div.search-item',
        listeners: {"render": function(combo) { Memoria.Search.InstantAdd.init(combo) },
    				"collapse": function() { Memoria.Search.InstantAdd.show() },
    				"select": function(combo, record) { 
    					POD.addTab(record.get("name"), "[% root %]/module/"+record.get("name"));
    					  Memoria.Search.InstantAdd.hide();
    					}, "blur": function(){Memoria.Search.InstantAdd.hide();} }
    

    });
    Ext.getDom('search').focus(100, true);
}

Ext.onReady(function(){
    Ext.Updater.defaults.loadScripts = true;
	
    var viewport = new Ext.Viewport({
        layout:'border',
        items:[POD.tabs, POD.leftColumn]
    });
});


lib/Catalyst/Controller/POD/POM/View.pm
package Catalyst::Controller::POD::POM::View;
use strict;
use warnings;

use base "Pod::POM::View::HTML";

my $HTML_PROTECT = 1;

sub _root {
    no strict 'refs';
    my $self = shift;
    if(@_) {
        $self->{_root} = $_[0];
    }
    return $self->{_root};
}


sub _module {
    no strict 'refs';
    my $self = shift;
    if(@_) {
        $self->{_module} = $_[0];
    }
    return $self->{_module};
}
sub _toc {
    no strict 'refs';
    my $self = shift;
    if(@_) {
        $self->{_toc} = $_[0];
    }
    return $self->{_toc};
}

sub view_begin {
    my ($self, $begin) = @_;
    return '' unless $begin->format() =~ /\bhtml\b/;
    $HTML_PROTECT = 1;
    my $output = $begin->content();
    return $output;
}


sub view_pod {
    my ($self, $pod) = @_;
    my $toc = $self->_toc;
    my $permalink = $self->_root . "?permalink=" . $self->_module; # There has to be a better way
    return qq~
<html>
<head>
<link rel="stylesheet" href="http://search.cpan.org/s/style.css" type="text/css" />

</head>
<body bgcolor=\"#ffffff\">
<script type="text/javascript">POD.setTOC($toc);</script>
<div id="permalink"><a href="$permalink">Permalink</a></div>
<div class="pod">
~
   . $pod->content->present($self)
        . "</div></body></html>\n";
}

sub view_verbatim {
    my ($self, $text) = @_;

    for ($text) {
        s/&/&amp;/g;
        s/</&lt;/g;
        s/>/&gt;/g;
    }
    foreach my $i (1..10) {
        my $t;
        my $last = 0;
        for(split(/\n/, $text)) {
            $_ =~ s/^(.)//;
            if($1 ne " ") {
                $last = 1;
                last;
            }
            $t .= $_.$/;
        }
        $text = $t unless($last);
        last if($last);
    }

    return "<pre>$text</pre>\n\n";
}

sub view_seq_link_transform_path {
    my($self,$page) = @_;
    return $self->_root."/module/$page";  
}
{ 
    no warnings 'redefine';
    sub Pod::POM::View::HTML::make_href  {
        my($url, $title) = @_;
    	if (!defined $url) {
            $url = "$title";
        }

        $title = $url unless defined $title;
        return qq{<a href="$url" onclick="return POD.proxyLink(this)">$title</a>};
    }
}

sub view_head1 {
    my ($self, $head1) = @_;
    my $title = $head1->title->present($self);
    my $id = "section-".$self->_module."-".$title;
    $id =~ s/<.*?>//g;
    $id =~ s/'/\\'/g;
    return "<h1 id='$id'>$title</h1>\n\n"
        . $head1->content->present($self);
}


sub view_head2 {
    my ($self, $head2) = @_;
    my $title = $head2->title->present($self);
    my $id = "section-".$self->_module."-".$title;
    $id =~ s/<.*?>//g;
    $id =~ s/'/\\'/g;
    return "<h2 id='$id'>$title</h2>\n\n"
        . $head2->content->present($self);
}


sub view_head3 {
    my ($self, $head3) = @_;
    my $title = $head3->title->present($self);
    my $id = "section-".$self->_module."-".$title;
    $id =~ s/<.*?>//g;
    $id =~ s/'/\\'/g;
    return "<h3 id='$id'>$title</h3>\n\n"
        . $head3->content->present($self);
}


sub view_head4 {
    my ($self, $head4) = @_;
    my $title = $head4->title->present($self);
    my $id = "section-".$self->_module."-".$title;
    $id =~ s/<.*?>//g;
    $id =~ s/'/\\'/g;
    return "<h4 id='$id'>$title</h4>\n\n"
        . $head4->content->present($self);
}

1;


lib/Catalyst/Controller/POD/Template.pm
package Catalyst::Controller::POD::Template;

use utf8;

sub get {
    my $class = shift;
    my $root = shift;
    return << "DATA"
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
    <head>
    <title>Pod::Browser</title>

    <link rel="stylesheet" type="text/css" href="$root/ext/resources/css/ext-all.css" />
    <link rel="stylesheet" href="$root/cpan.css" type="text/css" />
    <link rel="stylesheet" href="$root/docs.css" type="text/css" />
    <script type="text/javascript" src="$root/ext/adapter/ext/ext-base.js"></script>
    <script type="text/javascript" src="$root/ext/ext-all.js"></script>
    </head>
    <body>
        <script type="text/javascript">Ext.BLANK_IMAGE_URL = "$root/ext/resources/images/default/s.gif";</script>
        <script type="text/javascript" src="$root/docs.js"></script>
     </body>
</html>
DATA
}

1;


lib/Catalyst/Controller/POD.pm
package Catalyst::Controller::POD;
use warnings;
use strict;
use File::Find qw( find );
use File::ShareDir qw( dist_file );
use File::Spec;
use File::Slurp;
use Pod::Simple::Search;
use JSON::XS;
use Path::Class::File;
use Pod::POM;
use XML::Simple;
use LWP::Simple;
use List::MoreUtils qw(uniq);

our $VERSION = 0.02007;



use lib(qw(/Users/mo/Documents/workspace/Catalyst-Controller-POD/lib));
use Catalyst::Controller::POD::Template;

use base "Catalyst::Controller::POD::Search";

__PACKAGE__->mk_accessors(qw(_dist_dir inc namespaces self dir show_home_tab initial_module home_tab_content expanded_module_tree));

__PACKAGE__->config(
 self                 => 1,
 namespaces           => ["*"],
 initial_module       => "",
 show_home_tab        => 1,
 expanded_module_tree => 0,
 home_tab_content     => <<HTML,
<div style="width:500px; margin:50px" class='x-box-blue' id='move-me'>
<div class="x-box-tl"><div class="x-box-tr"><div class="x-box-tc"></div></div></div>
<div class="x-box-ml"><div class="x-box-mr"><div class="x-box-mc">
<h3 style="margin-bottom:5px;">Search the CPAN</h3>
<input type="text" name="search" id="search" class="x-form-text" style='font-size: 20px; height: 31px'/>
<div style="padding-top:4px;">Type at least three characters</div>
</div></div></div>
<div class="x-box-bl"><div class="x-box-br"><div class="x-box-bc"></div></div></div>
</div>
HTML
);

sub search : Local {
	my ( $self, $c ) = @_;
	my $k = $c->req->param("value");
	my $s = $c->req->param("start");
	my $url = new URI("http://search.cpan.org/search");
	$url->query_form_hash(
		query  => $k,
		mode   => "module",
		n      => 50,
		format => "xml",
		s      => $s
	);
	my $ua = new LWP::UserAgent;
	$ua->timeout(15);
	$c->log->debug("get url ".$url->canonical) if($c->debug);
	my $response = $ua->get($url);
	my $xml = $response->content;
	my $data;
	eval{ $data = XMLin($xml, keyattr => [] )};
	if(@$) {
		$c->res->body("[]");
		return;
	}
	my $output = {count => $data->{matches}};
	while(my($k,$v) = each %{$output->{module}}) {
		
	}
	$c->res->body(encode_json($data));
}


sub module : Local {
	my ( $self, $c, $module ) = @_;
	my $search = Pod::Simple::Search->new->inc( $self->inc || 0 );
	push( @{ $self->{dirs} }, $c->path_to('lib')->stringify )
	  if ( $self->{self} );
	my $name2path =
	  $search->limit_glob($module)->survey( @{ $self->{dirs} } );
	my $view = "Catalyst::Controller::POD::POM::View";
	Pod::POM->default_view($view);
	my $parser = Pod::POM->new( warn => 0 );
	$view->_root( $self->_root($c) );
	$view->_module($module);
	my $pom;

	if ( $name2path->{$module} ) {
		$c->log->debug("Getting POD from local store") if($c->debug);
		$view->_toc( _get_toc( $name2path->{$module} ) );
		$pom = $parser->parse_file( $name2path->{$module} )
		  || die $parser->error(), "\n";
	} else {
		$c->log->debug("Getting POD from CPAN") if($c->debug);
		my $html = get( "http://search.cpan.org/perldoc?" . $module );
	    my $source;
		if($html && $html =~ /.*<a href="(.*?)">Source<\/a>.*/) {
		    $html =~ s/.*<a href="(.*?)">Source<\/a>.*/$1/s;
    		$c->log->debug("Get source from http://search.cpan.org" . $html) if($c->debug);
    		$source = get( "http://search.cpan.org" . $html );
        } else {
            $source = "=head1 ERROR\n\nThis module could not be found.";
        }
		$view->_toc( _get_toc( $source ) );
		$pom = $parser->parse_text($source)
		  || die $parser->error(), "\n";
	}
	Pod::POM->default_view("Catalyst::Controller::POD::POM::View");
	$c->res->body( "$pom" );
}

sub _get_toc {
	my $source = shift;
	my $toc;
	my $parser = Pod::POM->new( warn => 0 );
	my $view = "Pod::POM::View::TOC";
	Pod::POM->default_view($view);
	my $pom = $parser->parse($source);
	$toc = $view->print($pom);
	return encode_json( _toc_to_json( [], split( /\n/, $toc ) ) );
}

sub _toc_to_json {
	my $tree     = shift;
	my @sections = @_;
	my @uniq     = uniq( map { ( split(/\t/) )[0] } @sections );
	foreach my $root (@uniq) {
		next unless ($root);
		push( @{$tree}, { text => $root } );
		my ( @children, $start );
		for (@sections) {
			if ( $_ =~ /^\Q$root\E$/ ) {
				$start = 1;
			} elsif ( $start && $_ =~ /^\t(.*)$/ ) {
				push( @children, $1 );
			} elsif ( $start && $_ =~ /^[^\t]+/ ) {
				last;
			}
		}
		unless (@children) {
			$tree->[-1]->{leaf} = \1;
			next;
		}
		$tree->[-1]->{children} = [];
		$tree->[-1]->{children} =
		  _toc_to_json( $tree->[-1]->{children}, @children );
	}
	return $tree;
}

sub modules : Local {
	my ( $self, $c, $find ) = @_;
	my $search = Pod::Simple::Search->new->inc( $self->{inc} || 0 );
	push( @{ $self->{dirs} }, $c->path_to('lib')->stringify )
	  if ( $self->{self} );
	my $name2path = {};

		for ( @{ $self->{namespaces} } ) {
			my $found =
			  Pod::Simple::Search->new->inc( $self->{inc} || 0 )
				  ->limit_glob($_)->survey( @{ $self->{dirs} } );
			%{$name2path} = (
				%{$name2path}, %{$found}
			);
		}
	
	my @modules;
	while ( my ( $k, $v ) = each %$name2path ) {
		next if($find && $k !~ /\Q$find\E/ig);
		push( @modules, $k );
	}
	@modules = sort @modules;
	my $json = _build_module_tree( [], "", @modules );
	$c->res->body( encode_json($json) );
}

sub _build_module_tree : Private {
	my $tree    = shift;
	my $stack   = shift;
	my @modules = @_;
	my @uniq    = uniq( map { ( split(/::/) )[0] } @modules );
	foreach my $root (@uniq) {
		my $name = $stack ? $stack . "::" . $root : $root;
		push( @{$tree}, { text => $root, name => $name } );
		my @children;
		for (@modules) {
			if ( $_ =~ /^$root\:\:(.*)$/ ) {
				push( @children, $1 );
			}
		}
		unless (@children) {
			$tree->[-1]->{leaf} = \1;
			next;
		}
		$tree->[-1]->{children} = [];
		$tree->[-1]->{children} =
		  _build_module_tree( $tree->[-1]->{children}, $name, @children );
	}
	return $tree;
}

sub _root {
	my ( $self, $c ) = @_;
	my $index = $c->uri_for( __PACKAGE__->config->{path} );

	#$index  =~ s/\/index//g;
	return $index;
}

sub new {
    my $class = shift;
    my $self  = $class->next::method(@_);
    my $file  = Path::Class::File->new( 'share', 'dist.js' );
    eval {
        $file = Path::Class::File->new(
            dist_file( 'Catalyst-Controller-POD', 'docs.js' ) );
    } unless(-e $file);
    $self->_dist_dir( $file->dir );
    return $self;
}


sub index : Path : Args(0) {
	my ( $self, $c ) = @_;
	$c->res->content_type('text/html; charset=utf-8');
	$c->response->body(
		Catalyst::Controller::POD::Template->get(
			$self->_root($c) . "/static"
		)
	);
}

sub get_home_tab_content : Path("home_tab_content") {
	my ( $self, $c ) = @_;
	$c->response->body($self->home_tab_content);
}

sub static : Path("static") {
	my ( $self, $c, @file ) = @_;
	my $file = File::Spec->catfile(@file);
	my $data;
	eval { $data = read_file( $self->_dist_dir . "/" . $file ) };
	if ($@) {
		$c->res->status(404);
		$c->res->content_type('text/html; charset=utf-8');
	} else {
		if ( $file eq "docs.js" ) {
			_replace_template_vars(\$data, "root",                       $self->_root($c));
			_replace_template_vars(\$data, "initial_module",             $self->initial_module);
			_replace_template_vars(\$data, "show_home_tab",              $self->show_home_tab ? "true" : "false");
			_replace_template_vars(\$data, "expand_module_tree_on_load", $self->expanded_module_tree ? "true" : "false");
		}
		$c->response->body($data);
	}
}

# A poor man's template module. 
sub _replace_template_vars {
	my ($data_ref, $var_name, $var_val) = @_;
	$$data_ref =~ s/\[% $var_name %\]/$var_val/g;
}

1;

__END__

=head1 NAME

Catalyst::Controller::POD - Serves PODs right from your Catalyst application

=head1 SYNOPSIS

Create a new controller and paste this code:

  package MyApp::Controller::YourNewController;  # <-- Change this to your controller
  
  use strict;
  use warnings;
  use base 'Catalyst::Controller::POD';
  __PACKAGE__->config(
    inc        => 1,
    namespaces => [qw(Catalyst::Manual*)],
    self       => 1,
    dirs       => [qw()]
  );
  1;

=head1 DESCRIPTION

This is a catalyst controller which serves PODs. It allows you to browse through your local
repository of modules. On the front page of this controller is a search box
which uses CPAN's xml interface to retrieve the results. If you click on one of them
the POD is displayed in this application.

Cross links in PODs are resolved and pop up as a new tab. If the module you clicked on is
not installed this controller fetches the source code from CPAN and creates the pod locally.
There is also a TOC which is always visible and scrolls the current POD to the selected section.

It is written using a JavaScript framework called ExtJS (L<http://www.extjs.com>) which
generate beautiful and intuitive interfaces.

Have a look at L<http://cpan.org/authors/id/P/PE/PERLER/pod-images/pod-encyclopedia-01.png>

=head1 CONFIGURATION

=over

=item dirs (Arrayref)

Search for modules in these directories.

Defaults to C<[]>.

=item expanded_module_tree (Boolean)

Expand the module browser tree on initial page load.

Defaults to C<1>

=item home_tab_content (String)

HTML to be displayed in the Home tab.

Defaults to the existing CPAN search box.

=item inc (Boolean)

Search for modules in @INC. Set it to 1 or 0.

Defaults to C<0>.

=item initial_module (String)

If this option is specified, a tab displaying the perldoc for the given module
will be opened on load.  Handy if you wish to disable the home tab and specify
a specific module's perldoc as the initial page a user sees.

Defaults to C<"">

=item namespaces (Arrayref)

Filter by namespaces. See L<Pod::Simple::Search> C<limit_glob> for syntax.

Defaults to C<["*"]>

=item self (Boolean)

Search for modules in C<< $c->path_to( 'lib' ) >>.

Defaults to C<1>.

=item show_home_tab (Boolean)

Show or hide the home tab.

Defaults to C<1>

=head1 NOTICE

This module works fine for most PODs but there are a few which do not get rendered properly. 
Please report any bug you find. See L</BUGS>.

Have a look at L<Pod::Browser> which is a catalyst application running this controller. You
can use it as a stand-alone POD server.

=head1 TODO

Write more tests!

=head1 AUTHOR

Moritz Onken <onken@netcubed.de>

=head1 CONTRIBUTORS

Tristan Pratt

=head1 BUGS

Please report any bugs or feature requests to C<bug-catalyst-controller-pod at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Catalyst-Controller-POD>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Catalyst::Controller::POD


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Catalyst-Controller-POD>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Catalyst-Controller-POD>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Catalyst-Controller-POD>

=item * Search CPAN

L<http://search.cpan.org/dist/Catalyst-Controller-POD>

=back


=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2008 Moritz Onken, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut